### ISSUE TITLE
Bug: Derivative of transposed-vector times matrix is incorrect.

### ISSUE THREAD

**Issue by Red-Portal** [JuliaDiff/ReverseDiff.jl#235]
Hi, 
Found a correctness bug that seems only to affect `ReverseDiff`.

```julia
julia> A = [1 2; 3 4]; x = [5, 6];

julia> [ReverseDiff.gradient(y -> sum(y'*A), x) ;; 
        ReverseDiff.gradient(y -> sum(A'*y), x) ;;
             Zygote.gradient(y -> sum(y'*A), x)[1] ;;
             Zygote.gradient(y -> sum(A'*y), x)[1] ]
2×4 Matrix{Float64}:
 4.0  3.0  3.0  3.0
 6.0  7.0  7.0  7.0
```

**Comment by mohamed82008** [comment]
Thanks for reporting this. Do you have any time to dig deeper into this?

**Comment by zelongyin2000** [comment]
Hi, I also found that the memory allocation benchmarks are no longer correct.

When I tried the example codes it now shows spurious memory allocations
![reversediff](https://github.com/JuliaDiff/ReverseDiff.jl/assets/123425254/53eb538f-93b3-4f62-abb8-0194cdad31e7)

It seems that these allocations are associated with the use of transposes. When I delete the adjoint operators in the definition they goes away.
